src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -440,139 +440,7 @@ public void OnHeader(Span<byte> name, Span<byte> value)
-                            // https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663
-                            //
-                            // This also prevents the 100 Continue response from being sent if the app
-                            // never tried to read the body.
-                            // https://github.com/aspnet/KestrelHttpServer/issues/2102
-                            //
-                            // ProduceEnd() must be called before _application.DisposeContext(), to ensure
-                            // HttpContext.Response.StatusCode is correctly set when
-                            // IHttpContextFactory.Dispose(HttpContext) is called.
-                            await ProduceEnd();
-
-                            // ForZeroContentLength does not complete the reader nor the writer
-                            if (!messageBody.IsEmpty)
-                            {
-                                // Finish reading the request body in case the app did not.
-                                await messageBody.ConsumeAsync();
-                            }
-                        }
-                        else if (!HasResponseStarted)
-                        {
-                            // If the request was aborted and no response was sent, there's no
-                            // meaningful status code to log.
-                            StatusCode = 0;
-                        }
-                    }
-                    catch (BadHttpRequestException ex)
-                    {
-                        // Handle BadHttpRequestException thrown during app execution or remaining message body consumption.
-                        // This has to be caught here so StatusCode is set properly before disposing the HttpContext
-                        // (DisposeContext logs StatusCode).
-                        SetBadRequestState(ex);
-                    }
-                    finally
-                    {
-                        application.DisposeContext(httpContext, _applicationException);
-
-                        // StopStreams should be called before the end of the "if (!_requestProcessingStopping)" block
-                        // to ensure InitializeStreams has been called.
-                        StopStreams();
-
-                        if (HasStartedConsumingRequestBody)
-                        {
-                            RequestBodyPipe.Reader.Complete();
-
-                            // Wait for MessageBody.PumpAsync() to call RequestBodyPipe.Writer.Complete().
-                            await messageBody.StopAsync();
-
-                            // At this point both the request body pipe reader and writer should be completed.
-                            RequestBodyPipe.Reset();
-                        }
-                    }
-                }
+                await ProcessRequests(application);
             }
             catch (BadHttpRequestException ex)
```

Comment:
Can this still happen? At this level all error are connection oriented right?

Reply:
Parsing headers and request line can throw a `BadHttpRequestException` which gets caught here

Reply:
I see

Reply:
Is before any request has started so it catches here and tears down the connection